[DO NOT MERGE!!!] Introduce tilesPerWarp parameter, scale preshuffling and bypassingLDS

plognjen opened the pull request.

zhanglx13 commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -674,6 +654,8 @@ LinearLayout mfmaDotToLinearLayout(DotOperandEncodingAttr dotMfmaLayout,
   int32_t kSize = shape[kDim];
   auto warpsPerCTA = mfmaLayout.getWarpsPerCTA();
+  auto tilesPerWarp = mfmaLayout.getTilesPerWarp();
+  auto tilePerWarpNonK = tilesPerWarp[kDim];
```
this `tilesPerWarp` is for mfmaLayout, which does not have "kDim". Can you be more specific here?

zhanglx13 commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -1460,31 +1451,16 @@ LinearLayout chooseScaledMfmaScaleLayout(
-  auto regs = mlir::triton::identityStandardND(kRegister, {1, 1}, order);
-  LinearLayout lanes = LinearLayout::empty();
+  auto kDim = dotOperandIdx == 0 ? rank - 1 : rank - 2;
+  auto tilePerWarpNonK = tilesPerWarp[kDim];
```
here we are dealing with operands and there is kDim and nonKDim. If you need nonKDim, you should use tilesPerWarp[nonKDim]

zhanglx13 commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -1494,43 +1470,70 @@ LinearLayout chooseScaledMfmaScaleLayout(
-                      {kBlock, {}}},
-                     {standardOutDims[order[0]], standardOutDims[order[1]]});
+    if (preshuffleScales) {
+      laneBase = {{4, 0}, {0, 1}, {0, 2}, {0, 4}, {0, 8}, {0, 16}};
```
Can you add doc for this LL? It looks like mfma32 scale layout?

oplavsic committed:
[AMD] Introduce tilesPerWarp and scale preshuffling

zhanglx13 committed:
Refactor

- Only swap elems in getValuesFromDotOperandLayoutStruct if preshuffle
env var is set
- Only set tilesPerWarp = [2, 2] for scaledDot with mxfp4 case

oplavsic committed:
Generalize preshuffling in dot lowering for any block sizes
